### PR TITLE
Product management

### PR DESCRIPTION
**Pull request by FAEZ10**
# **Pull Request: Product Management Enhancements**

## **Type of Change**
- [x] Feature
- [ ] Fix

---

## **Description**

This PR introduces multiple enhancements and new features for product and product-related functionalities, including CRUD operations, advanced methods, and service integrations. Below is a summary of the changes:

### **1. Product Management Enhancements**
- Implemented `toggleProductActivation` to activate or deactivate products.
- Added `searchProducts` for advanced filtering based on name, categories, and price range.
- Introduced `duplicateProduct` to create a copy of an existing product.
- Added `getProductsByStatus` to filter products by their activation status.
- Developed `addImagesToProduct` to associate images with products.
- Created `removeImageFromProduct` to dissociate and delete images from products.

### **2. Product Variant Management**
- Implemented CRUD operations for managing product variants.
- Added `searchVariants` for advanced filtering based on product, size, color, and price range.
- Introduced `getLimitedEditionVariants` to retrieve variants marked as limited editions.
- Created `duplicateVariant` to clone an existing product variant.

### **3. Product Features and Categories**
- Developed `ProductHasCategoryService`:
  - `linkProductToCategory` to associate a product with a category.
  - `unlinkProductFromCategory` to remove associations between products and categories.
- Developed `ProductHasProductFeatureService`:
  - `addFeatureToProduct` to associate features with products and assign values.
  - `removeFeatureFromProduct` to dissociate features from products.

### **4. Product Images**
- Added CRUD operations for managing product images.
- Integrated routes for creating, updating, retrieving, and deleting images.
- Enhanced validation to ensure consistency in image-product relationships.

---

## **How Has This Been Tested?**

- [x] Manually tested all API endpoints using Postman.
- [x] Verified relationships between products, categories, and features.
- [x] Ensured proper association and dissociation of images with products.
- [x] Tested all CRUD operations for Product and ProductVariant services.
- [x] Validated advanced search and filtering functionalities.

---

## **Checklist**
- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have made corresponding changes to the documentation (Swagger).
- [x] My changes generate no new warnings.
- [x] I have added tests that prove my feature works.
- [x] New and existing unit tests pass locally with my changes.
- [x] Any dependent changes have been merged and published in downstream modules.

---

## **Relevant Issues or Context**

This PR consolidates the development of features for the **`product-management`** branch, including advanced methods and CRUD operations for the following:
- Products
- Product Variants
- Product Features
- Product Categories
- Product Images

### CODE REVIEW

**Review by johnnyhelloworld — APPROVED** [review]
Review code + Testes des routes sur Postman OK

File: src/apps/product/api/product.route.ts [hunk 1]
```diff
@@ -69,7 +73,45 @@ router.post(
...
+ *       - in: query
+ *         name: maxPrice
+ *         schema:
+ *           type: number
+ *         description: Maximum price
+ *     responses:
+ *       200:
+ *         description: Products retrieved successfully
+ */
+router.get("/search", authenticateMiddleware, productController.searchProducts.bind(productController));
```
Comment (DaniOmer) [inline src/apps/product/api/product.route.ts:113]: Je pense qu'un user non connecté devrait pouvoir rechercher un produit

File: src/apps/product/api/product.route.ts [hunk 2]
```diff
@@ -96,6 +138,7 @@ router.get("/", productController.getAllProducts.bind(productController));
  */
 router.get(
   "/:id",
+  authenticateMiddleware,
   validateIdMiddleware("Product"),
   productController.getProductById.bind(productController)
 );
```
Comment (DaniOmer) [inline src/apps/product/api/product.route.ts:144]: Pareil ici, normal on peut afficher un produit sans être connecté

File: src/apps/product/api/productFeature/productFeature.route.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  "/",
+  authenticateMiddleware,
+  checkRoleMiddleware(["admin"]),
+  validateDtoMiddleware(ProductFeatureCreateDTO),
+  productFeatureController.create.bind(productFeatureController)
+);
+
+router.get("/",
+    authenticateMiddleware,
+   productFeatureController.getAll.bind(productFeatureController));
```
Comment (DaniOmer) [inline src/apps/product/api/productFeature/productFeature.route.ts:31]: Idem

File: src/apps/product/data-access/product.model.ts [hunk 4]
```diff
@@ -1,10 +1,20 @@
...
+  {
     name: { type: String, required: true },
-    description: { type: String, default: null },
-    }, { timestamps: true, collection: "products", versionKey: false
-  });
-  
-export const ProductModel = model<IProduct>("Product", ProductSchema);
\ No newline at end of file
+    description: { type: String },
+    basePrice: { type: Number },
```
Comment (DaniOmer) [inline src/apps/product/data-access/product.model.ts:8]: Ici il faut qu'on ait la même chose que dans product variant `priceEtx` et `priceVat`

File: src/apps/product/domain/product.service.ts [hunk 5]
```diff
@@ -35,4 +69,169 @@ export class ProductService extends BaseService {
...
+        context: { productId: `No product found with ID: ${id}` },
+        code: 404,
+      });
+    }
+
+    const duplicatedProduct = {
+      ...product.toObject(),
+      _id: undefined,
+      name: `${product.name} (Copy)`,
+    };
```
Comment (DaniOmer) [inline src/apps/product/domain/product.service.ts:120]: je ne sais pas si c'est pertinent le duplicate product. Mais niveau code pour faire plus propre tu devrais destructurer autrement, récupérer les champs que tu peux de product sans prendre id, _id, createdAt et updatedAt

File: src/apps/product/domain/productVariant/productVariant.service.ts [hunk 6]
```diff
@@ -34,4 +47,87 @@ export class ProductVariantService extends BaseService {
...
+
+  async duplicateVariant(id: string): Promise<IProductVariant> {
+    const variant = await this.repository.getById(id);
+    this.validateDataExists(variant, id);
+
+    const duplicatedVariantData = {
+      ...variant?.toObject(),
+      _id: undefined,
+      stripeId: undefined, 
+    };
```
Comment (DaniOmer) [inline src/apps/product/domain/productVariant/productVariant.service.ts:105]: idem que pour product

**Pull request closed without merging**